### PR TITLE
Block transfer support

### PR DESCRIPTION
**Pull request by tuve**
Full block transfer support including on both client and server including:
- block1 registrations
- block1 write requests
- block2 read responses
- block2 observe responses
- block transfer retransmission on entity too large response

Signed-off-by: Tuve Nordius <tuve.nordius@husqvarnagroup.com>

### CODE REVIEW

File: core/liblwm2m.h
Comment (sbertin-telular) [inline core/liblwm2m.h]: Minor typo, migth should be might.

File: examples/client/test_object.c
Comment (sbertin-telular) [inline examples/client/test_object.c]: Leshan lists this as optional, not mandatory. See https://github.com/eclipse/leshan/blob/master/leshan-server-demo/src/main/resources/models/31024.xml.

File: examples/client/test_object.c
Comment (sbertin-telular) [inline examples/client/test_object.c]: Should make sure it is a string or opaque type first or dataArray[i].value.asBuffer.length may not be valid.

File: core/registration.c
Comment (AButenko) [inline core/registration.c:838]: This will lead to double free() call in this implementation

File: core/registration.c
Comment (AButenko) [inline core/registration.c:905]: Here in case of the registration update with objects included, we will free the payload, but not set it to NULL in the dataP. This leads to double-free in prv_handleRegistrationUpdateReply

**Review by AButenko — CHANGES_REQUESTED** [review]

File: core/liblwm2m.h
Comment (tuve) [reply core/liblwm2m.h]: Fixed

File: examples/client/test_object.c
Comment (tuve) [reply examples/client/test_object.c]: I have changed it resource id 5 to be non conflicting with the leshan object spec, but now I can't find it at all in the leshan project

File: examples/client/test_object.c
Comment (tuve) [reply examples/client/test_object.c]: Fixed

File: core/registration.c
Comment (tuve) [reply core/registration.c:905]: Thank you, I currently looking into some other (possible) memory issues and will have a look at this as well.

File: core/block.c
Comment (stefanfangvall) [inline core/block.c:62]: Is it guaranteed that the identifier URI's are always \0 terminated, otherwise use strncmp()

Also see comment in liblwm2m.h

File: core/block.c
Comment (stefanfangvall) [inline core/block.c:235]: No check that malloc returned != NULL, hardfault would occur in memcpy()

File: core/er-coap-13/er-coap-13.c
Comment (stefanfangvall) [inline core/er-coap-13/er-coap-13.c:397]: Remove?

File: core/er-coap-13/er-coap-13.c
Comment (stefanfangvall) [inline core/er-coap-13/er-coap-13.c:401]: Check if output is != NULL

File: core/er-coap-13/er-coap-13.c
Comment (stefanfangvall) [inline core/er-coap-13/er-coap-13.c:780]: Paranthesis

File: core/liblwm2m.h
Comment (stefanfangvall) [inline core/liblwm2m.h:522]: Rename, all other uri members are of type lwm2m_uri_t

File: core/observe.c
Comment (stefanfangvall) [inline core/observe.c:880]: Calling callbacks without checking arguments is not nice, who knows what the callback does...

File: core/observe.c
Comment (stefanfangvall) [inline core/observe.c:888]: Calling callbacks without checking arguments is not nice, who knows what the callback does...

File: core/observe.c
Comment (stefanfangvall) [inline core/observe.c:901]: Calling callbacks without checking arguments is not nice, who knows what the callback does...

File: core/observe.c
Comment (stefanfangvall) [inline core/observe.c:909]: Calling callbacks without checking arguments is not nice, who knows what the callback does...

File: core/observe.c
Comment (stefanfangvall) [inline core/observe.c:1014]: Calling callbacks without checking arguments is not nice, who knows what the callback does...

File: core/observe.c
Comment (stefanfangvall) [inline core/observe.c:1023]: Calling callbacks without checking arguments is not nice, who knows what the callback does...

File: core/registration.c
Comment (stefanfangvall) [inline core/registration.c:1580]: Parenthesis

File: core/transaction.c
Comment (stefanfangvall) [inline core/transaction.c:499]: Check context == NULL

File: core/utils.c
Comment (stefanfangvall) [inline core/utils.c:559]: Check contextP == NULL

**Comment by qleisan** [comment]
squashed and rebased to **qleisan/wakaama:blocktransfer_squash_rebase** (to join the effort)

**Comment by qleisan** [comment]
A rebased (to leatest-greatest) version of husqvarnagroup:blocktransfer (this PR) is now available at **husqvarnagroup:blocktransfer2** (IMHO the most pragmatic way to coordinate work, avoid git complications)

Serveral issues have been created. My suggestion on the way forward so that we can merge this "major" PR)

#510 #511 should be resolved on blocktransfer2 ( @tuve @qleisan )
no other issues need to be resolved before merge - _**any other opinions?**_
the commits on blocktransfer2 are squashed and signed. A new PR is submitted and this is PR closed

Sounds like a plan?

**Comment by sbernard31** [comment]
> husqvarnagroup:blocktransfer2 (IMHO the most pragmatic way to coordinate work, avoid git complications)

If you want to keep **blocktransfer**,  you just checkout blocktransfer branch, then reset --hard on you branch blocktransfer2, then push force blocktransfer and this PR should be updated as expected.

**Comment by qleisan** [comment]
I read this ["When pushing commits to a pull request, don't force push. Force pushing can corrupt your pull"](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/about-pull-requests)
 and generally felt is was safer to keep history. But I can do what you suggest if nobody minds overwriting the old branch...

**Comment by sbernard31** [comment]
> I read this "When pushing commits to a pull request, don't force push. Force pushing can corrupt your pull"

I'm surprised about that :open_mouth: ! Personally, I do push force all the time in my PR with github.
I tried to better understand reading : https://stackoverflow.com/questions/50132325/github-advises-force-pushing-can-corrupt-your-pull-request-why and I don't get the issue :thinking: 

> and generally felt is was safer to keep history. But I can do what you suggest if nobody minds overwriting the old branch...

I have no strong opinion, I just give you some steps to help you in case you want to keep this PR. :slightly_smiling_face: 

I consider rebase better than merge, mainly because I think to keeping repository readable is more important than keeping history.
And in this case, if we don't want to use push force that means that each rebase will need a new PR, not sure this is want we want ?

(Previously @sbertin-telular just rebase + push --force  its PR)

**Comment by sbertin-telular** [comment]
The problem with force pushing comes if someone has pulled down that branch and started making changes against it. That can cause issues. A rebase or editing a commit message to add a forgotten signoff will require a force push or a completely new pull request. I'm OK with a force push in this case.

**Comment by qleisan** [comment]
> If you want to keep blocktransfer, you just checkout blocktransfer branch, then reset --hard on you branch blocktransfer2, then push force blocktransfer and this PR should be updated as expected.

Done! What could possibly go wrong ;-) 
#511 partially fixed (no compilation warnings or crash), but the test needs more work. Imho no longer a merge blocker

@tuve if you do work on #510 add it to (this) "blocktransfer" branch. (I saved a "copy" before I moved it as "blocktransfer_backup")

**Comment by tuve** [comment]
I think we can merge this, I am still working on #510, but maybe in the interest of moving forward and not blocking other work, we could do separate pull request for that and the relevant comments from @stefanfangvall. What do you think @sbernard31 and @sbertin-telular ?

**Comment by sbernard31** [comment]
On my side, I have no preference. So I'm ok to integrate this one if wanted. (I'm waiting for @sbertin-telular opinion before)

About "not blocking other work", there is 3 remaining PRs.
This one which sounds to be in mergeable state. 
#433, I'm not sure I get what we should do with it ?
And last but not least the code reorg #431 which could be impacting. 

Maybe the idea would be to integrate #431 before to invest too much effort in new work to avoid too much conflicts ? :thinking:

**Comment by sbertin-telular** [comment]
I'd like to see a solution to #510 before merging even if it needs to later change based on feedback from IETF. I think interoperability with Leshan is important. @tuve are we talking days, weeks, or months to have a solution for #510. I think if it would be done by January we'll be good. With the Christmas holiday coming up  I know there are other demands on my time at least that might mean little progress before then, so not much blocked by waiting.

For #433, I think we can close it without merging if we don't hear back from the author.
#431 could impact future changes and I was planning to work on that next before merging more of my LWM2M 1.1 changes.

**Comment by tuve** [comment]
We are talking days, it is not many lines of code .... just need to make sure I get it right.

**Comment by sbertin-telular** [comment]
@tuve, what is the status of this? I have a rework of #431 almost ready but would like to give you an opportunity to finish this first.

**Comment by tuve** [comment]
> @tuve, what is the status of this? I have a rework of #431 almost ready but would like to give you an opportunity to finish this first.

Sorry for the late reply. I think I have everything in place that the pull request could be finished, just need to sort out git.

**Comment by tuve** [comment]
@sbernard31 and @sbertin-telular  this is ready to merge

**Comment by sbernard31** [comment]
@sbertin-telular  do you want to review it before I integrate it ? 
@tuve is there someone else who already review the code on your side ? (maybe @qleisan  or @rettichschnidi)

Should I squash both commit or you prefer to keep it separated ?

**Comment by tuve** [comment]
You can squash them

File: core/block.c
Comment (tuve) [reply core/block.c:230]: Fixed

File: core/er-coap-13/er-coap-13.c
Comment (tuve) [reply core/er-coap-13/er-coap-13.c:397]: Fixed

File: core/er-coap-13/er-coap-13.c
Comment (tuve) [reply core/er-coap-13/er-coap-13.c:401]: Fixed

File: core/er-coap-13/er-coap-13.c
Comment (tuve) [reply core/er-coap-13/er-coap-13.c:780]: It is an assign and my linter (and possible compiler) does complain if I remove the parenthis, no changes.

File: core/liblwm2m.h
Comment (tuve) [reply core/liblwm2m.h:537]: The char * is a used as an unique identifier constructed from both the uri and query parameters, there is no need to handle the uri or query parameters as structs and will in fact make it more complicated in this case. No changes

File: core/observe.c
Comment (tuve) [reply core/observe.c:896]: I think we had similar discussions in other cases and the argument beeing that in order to keep a minimal footprint we need to trust calling functions and callback to not do anything vile.

File: core/observe.c
Comment (tuve) [reply core/observe.c:904]: I think we had similar discussions in other cases and the argument beeing that in order to keep a minimal footprint we need to trust calling functions and callback to not do anything vile.

File: core/observe.c
Comment (tuve) [reply core/observe.c:917]: I think we had similar discussions in other cases and the argument beeing that in order to keep a minimal footprint we need to trust calling functions and callback to not do anything vile.

File: core/observe.c
Comment (tuve) [reply core/observe.c:925]: I think we had similar discussions in other cases and the argument beeing that in order to keep a minimal footprint we need to trust calling functions and callback to not do anything vile.

File: core/observe.c
Comment (tuve) [reply core/observe.c:1030]: I think we had similar discussions in other cases and the argument beeing that in order to keep a minimal footprint we need to trust calling functions and callback to not do anything vile.

File: core/observe.c
Comment (tuve) [reply core/observe.c:1039]: I think we had similar discussions in other cases and the argument beeing that in order to keep a minimal footprint we need to trust calling functions and callback to not do anything vile.

File: core/registration.c
Comment (tuve) [reply core/registration.c:1583]: Does || has precedence over && ? I think it is the otherway around. Regardless, this makes it clearer. No changes.

File: core/transaction.c
Comment (tuve) [reply core/transaction.c:495]: I think we had similar discussions in other cases and the argument beeing that in order to keep a minimal footprint we need to trust calling functions and callback to not do anything vile.

File: core/utils.c
Comment (tuve) [reply core/utils.c:937]: I think we had similar discussions in other cases and the argument beeing that in order to keep a minimal footprint we need to trust calling functions and callback to not do anything vile.

**Comment by tuve** [comment]
@sbernard31 I realized that I hadn't addressed the comments by @stefanfangvall. Updated with fixes where appropriate.

We have also done quite a lot of testing on a device, but it would still be nice if someone else could have a look at it to (@sbertin-telular , @qleisan)

**Comment by boaks** [comment]
Hi together,

(I'm not sure, if there is an issue about the blockwise and it may be better to comment there, anyway ...)
 
I spent recently some time in the blockwise in Californium. Though I stumbled over some pain in changing the blocksize, I sent a question to the IETF core mailing list and got surprising answer (at least it was surprising to me :-) ) :

[blocks matching](https://mailarchive.ietf.org/arch/msg/core/z9_HsDxAQJ17cqFwz2QhViOsZDI/)

So, using the block-num for block-matching seems to be wrong (or, may at least cause issues). The matching is better done on the offset (block-num * size).

**Comment by sbernard31** [comment]
I didn't read the code but I played a little with Wakaama client example and Leshan server demo.
I'm able to read and write some data on 31024/10/5 with SenML-JSON out of the box (without any Leshan configuration).
Looking at the wireshark capture, I didn't see anything wrong.

Just 1 strange behavior :thinking: which looks like last issue reported at https://github.com/eclipse/wakaama/issues/510#issue-759570278.  
I write a payload on `31024/10/5` then I read it : 
- Tested with 432KB :  write  :heavy_check_mark:  read :heavy_check_mark: 
- Tested with 595KB :  write :heavy_check_mark:  read :x: (INTERNAL_SERVER_ERROR)

**Comment by tuve** [comment]
With SenML-JSON I can replicate it as well, the problem seems not to be the blocktransfer but rather the senml_json_serialize function and more specific PRV_JSON_BUFFER_SIZE if that is increased it will work.

**Comment by tuve** [comment]
@boaks Thank you for comment. I think we use block_num * size for the offset since it is perfectly legal to "skip" block_nums if changing to a smaller size. At least this is the intention, if it is not done like that it is a misstake. As with 4.13 entity to large I believe we do it in the same way as it is done in californium, that was a surprise to me too and we had a long discussion about it in #510.

**Comment by boaks** [comment]
My comment was more about request/response matching, e.g. request block n1/s1 and get a response with n2/s2, then they match with `n1*s1` == `n2*s2` and not always with `n1` == `n2`.

**Comment by qleisan** [comment]
> With SenML-JSON I can replicate it as well, the problem seems not to be the blocktransfer but rather the senml_json_serialize function and more specific PRV_JSON_BUFFER_SIZE if that is increased it will work.

I can confirm Tuves finding (works fine with TLV). Should not be a blocker for this PR.

**Comment by qleisan** [comment]
> We have also done quite a lot of testing on a device, but it would still be nice if someone else could have a look at it to (@sbertin-telular , @qleisan)

Comments about potential runtime errors (not checking if malloc returns null pointer etc.) have been fixed. The issues I found while testing #510 have been addressed. I'm happy

**Comment by sbertin-telular** [comment]
I'll accept the approval of @qleisan for merging this. I haven't been able to find time to thoroughly review it myself.

**Comment by sbernard31** [comment]
This is integrated in master (commit d6b6b2a741ba08e2330298b2fa10d90d9117fc7b)

Thx guys :+1:.

**Pull request closed without merging**
